### PR TITLE
PR for presentation

### PR DESCRIPTION
**Pull request by kathryn-willi**
I started the code for making the flag plots size- and color-coded in your presentation.Rmd
I added my work on the window flag anomaly function -- it does not work yet. Please disregard!
I added a function were I mined the field notes to find known instances of a sensor malfunction; so far, it is just for 2022.
I added some new summary data associated with the slope. This should not change the output of the workflow, but I'm wanting to delve a bit more into additional slope-specific flags.

### CODE REVIEW

File: src/qaqc/download_and_flag_fxns/add_seasonal_flag.R [hunk 1]
```diff
@@ -23,7 +23,12 @@ add_seasonal_flag <- function(df) {
     # flag obs whose slope is greater than the 99th percentile range
     add_flag((slope_ahead >= t_slope_behind_99 | slope_behind >= t_slope_behind_99), "slope violation") %>%
     # ... flag obs that are outside of the rolling mean times 3 sd's of the 1-99 percentile seasonal values
-    add_flag((mean <= rollavg - (3 * t_sd_0199) | mean >= rollavg + (3 * t_sd_0199)), "outside sd range")
+    add_flag((mean <= rollavg - (3 * t_sd_0199) | mean >= rollavg + (3 * t_sd_0199)), "outside sd range") %>%
+    # ... flag obs that are outside of the rolling mean times 3 sd's of the 1-99 percentile seasonal values
+    add_flag((mean <= rollavg - (3 * t_sd_0199) | mean >= rollavg + (3 * t_sd_0199)), "outside sd range") #%>%
```
Comment (juandlt-csu) [inline src/qaqc/download_and_flag_fxns/add_seasonal_flag.R:28]: It looks like this flag is being added twice.

File: src/qaqc/download_and_flag_fxns/generate_summary_statistics.R [hunk 2]
```diff
@@ -32,6 +33,9 @@ generate_summary_statistics <- function(site_param_df) {
       # Determine the slope of a point in relation to the point ahead and behind.
       slope_ahead = abs(front1 - mean)/15,
       slope_behind = abs(mean - back1)/15,
+      rollslope = roll_mean(slope_behind, n = 7, align = 'center', na.rm = F, fill = NA_real_),
+      # Add the standard deviation for points centered in a rolling slope of 7 points.
```
Comment (juandlt-csu) [inline src/qaqc/download_and_flag_fxns/generate_summary_statistics.R:37]: question: does this mean that we are looking for the sd inside of a rolling 7 point period?

File: src/qaqc/download_and_flag_fxns/kw_add_large_anomaly_flag.R [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  # only "sonde not employed", "missing data", and "anomaly window"
+  # only "site visit", "missing data", "24hr anomaly", and "anomaly window"
+  # only "sv window", "missing data", "24hr anomaly", and "anomaly window"
+  # flag_string <- "^(24hr anomaly|anomaly window|24hr anomaly;\\nanomaly window|
+  #                 missing data|sonde not employed;\\nmissing data|
+  #                 missing data;\\nanomaly window|missing data;\\n24hr anomaly|
+  #                 sonde not employed;\nmissing data;\nanomaly window|
+  #                 missing data;\n24hr anomaly;\nanomaly window|
+  #                 site visit;\nmissing data;\n24hr anomaly;\nanomaly window|
+  #                 sv window;\nmissing data;\n24hr anomaly;\nanomaly window)$"
```
Comment (juandlt-csu) [inline src/qaqc/download_and_flag_fxns/kw_add_large_anomaly_flag.R:29]: Question: how does the new flag string work? I see that there is "missing data" inside of it twice.

File: src/qaqc/download_and_flag_fxns/kw_add_large_anomaly_flag.R [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  # Define a function to check if a given 6-hour window has >= 50% fails
+  check_6_hour_window_fail <- function(x) {
+    sum(x) / length(x) >= 0.5
+  }
+
+  df <- df %>%
+    mutate(flag_binary = ifelse((is.na(flag) | grepl(flag_string, flag)), 0, 1)) %>%
+    #arrange(timestamp) %>%
+    mutate(over_50_percent_fail_window = zoo::rollapply(flag_binary, width = 24, FUN = check_6_hour_window_fail, fill = NA, align = "right")) %>%
+    add_flag(over_50_percent_fail_window == TRUE, "anomaly window")
```
Comment (juandlt-csu) [inline src/qaqc/download_and_flag_fxns/kw_add_large_anomaly_flag.R:42]: Praise: this is way easier to read. Does the align = "right" mean that the window is not centered in the middle and instead at the end?

File: src/qaqc/download_and_flag_fxns/kw_known_probs.R
Comment (juandlt-csu) [inline src/qaqc/download_and_flag_fxns/kw_known_probs.R:1]: Question: should the sensor malfunction flag lead to the data points automatically failing for all the data points in that date time or should it be just for that sensor?

File: src/qaqc/02_explore_and_fix.Rmd [hunk 5]
```diff
@@ -97,3 +95,144 @@ Histograms of all parameter/site combos
 #   ggsave(filename, plot, device = "png", width = 8, height = 6)
 # })
 ```
+
+# Is the seasonal threshold flag doing anything useful?
```
Comment (juandlt-csu) [inline src/qaqc/02_explore_and_fix.Rmd:99]: Thoughts?

File: .gitignore [hunk 6]
```diff
@@ -3,24 +3,16 @@
 .RData
 .Ruserdata
 .DS_Store
+data
```
Comment (juandlt-csu) [inline .gitignore:6]: Thank you for cleaning this up :)

**Review by juandlt-csu — APPROVED** [review]
This looks good to merge to me!

File: src/qaqc/download_and_flag_fxns/kw_known_probs.R
Comment (kathryn-willi) [reply src/qaqc/download_and_flag_fxns/kw_known_probs.R:1]: Just for the sensor(s) listed in the sensor malfunction argument.

File: src/qaqc/download_and_flag_fxns/add_seasonal_flag.R (hunk 1 above)
Comment (kathryn-willi) [reply src/qaqc/download_and_flag_fxns/add_seasonal_flag.R:28]: good catch!

File: src/qaqc/download_and_flag_fxns/generate_summary_statistics.R (hunk 2 above)
Comment (kathryn-willi) [reply src/qaqc/download_and_flag_fxns/generate_summary_statistics.R:37]: yes! I'm exploring ways we might be able to determine sensor drift, especially instances where turbidity increases at a steady degree due to build-up. My idea is that we could identify instances where - if slope remains constant (give or take some threshold of difference) over a certain number of days, we flag it... still a mental work in progress, for sure!

File: src/qaqc/download_and_flag_fxns/kw_add_large_anomaly_flag.R (hunk 3 above)
Comment (kathryn-willi) [reply src/qaqc/download_and_flag_fxns/kw_add_large_anomaly_flag.R:29]: You can disregard all of this script. I will move it to "scratch".

**Pull request merged**